README.md
# To Run this Project

- You need to have `Go v1.17` or superior installed for the back end
- Clone this project
- Make sure you have a Local MongoDB database set up on port 27017 (default port)
- To execute the back end of the app, run `go run app/main.go`
- On Postman execute the following cURL: `curl --location --request POST 'http://localhost:3000/v1/preload'` to populate the DB with the required Users and Chatrooms data (you should see 3 users under Users collection, and 2 chatrooms under Chatrooms collection)
- Get the Docker project for the front available at [DockerHub](https://hub.docker.com/r/ratolibre1/front). To do so, run `docker pull ratolibre1/front` followed by `docker run -p 4200:4200 ratolibre1/front`

# To Review this Project

- Load `http://localhost:4200` on two browsers
- You should see the Login section
- Use credentials `email: artichoke@jobsity.com, password: 123456` on one of them and `email: lilbasil@jobsity.com, password: 123456` on the other (wrong email and/or password will result in failure to login)
- The Chatrooms section should appear
- Enter the same chatroom with both users. When you type a message with one of them it will show on both chats (unless it's a command)
- Chat autoscrolls when a new message is received (most of the time, it doesn't behave too well with the stock responses)
- You can change chatrooms
- If you send a valid stock command (like `/stock=aapl.us`) the message won't be stored but a response will be generated and stored

# Some considerations

- I tried to complete all the mandatory points. I got stuck implementing Unit Tests as it's something I've never done before in Golang (I can write UTs in Angular just fine)
- I chose hexagonal architecture as I believe it is very orderly and allows devs to separate the layers of the app
- I tried to work as if I was part of a team doing Scrum. Instead of Jira I kept a [Trello Board](https://trello.com/invite/b/sgfmFApA/ATTI4dd11900533d406bf88694ea4daa3309FCC10A6A/chat-hex) with my list of tasks.
- I implemented JWT for Authentication as they are simple enough to implement and secure enough to block access to protected URLs
- The RabbitMQ code is a bit sloppy as I was focusing more in getting it done than in doing it nice. Apologies for that, it still gets the job done
- I wanted to implement websockets for the chat to work in real time. I didn't have the time to so instead the front just pings the messages endpoint once per second as an MVP
- The front is a bit ugly but I expect it to be understandable enoguh. I promise I can write more beautiful Angular code if given the time
  
- You might wonder the meaning behind the two test emails. We have two wonderful calico cats called Alcachofa ("Artichoke" in Spanish) and Albahaquita ("Little Basil" in Spanish). You can see them both [here](https://photos.app.goo.gl/aNXfAAouLwFQPUag7) (Chofi is the black headed one, Baqui is the white headed one)
- Thank you for the opportunity! I hope this project showcases my skills and what I can bring to the team in terms of coding


CHANGELOG
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---

## [0.0.12] - 2023-10-18 (CHB-5)

### Added

- Added emitter and listener to use RabbitMQ for processing of commands by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.11] - 2023-10-18 (CHB-14)

### Added

- Added endpoint to preload the needed data in mongoDB by [Arturo Vial](https://github.com/ratolibre1)

### Removed

- Removed unnecessary files from the old base project ("user" files) by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.10] - 2023-10-18 (CHB-4)

### Added

- Added connection to Stock endpoint checking valid response and storing it in DB if valid by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.9] - 2023-10-18 (CHB-12)

### Changed

- List of messages is now returned in the right order by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.8] - 2023-10-17 (CHB-11)

### Added

- Added CORS permissions so local Angular can connect to local Go by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.7] - 2023-10-17 (CHB-9)

### Added

- Added new endpoint to get full list of chatrooms by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.6] - 2023-10-17 (CHB-7)

### Added

- Added new endpoints to validate login details and return a JWT token by [Arturo Vial](https://github.com/ratolibre1)
- Added Auth module by [Arturo Vial](https://github.com/ratolibre1)
- Added functions to find users by email, hash and validate passwords by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.5] - 2023-10-16 (CHB-6)

### Added

- Added new endpoints to enter and leave chatrooms by [Arturo Vial](https://github.com/ratolibre1)
- Added Chatrooms and Users modules by [Arturo Vial](https://github.com/ratolibre1)

### Changed

- Cleaned some English errors in original project by [Arturo Vial](https://github.com/ratolibre1)

### Removed

- Removed some calls to original User package (to be removed in the future) by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.4] - 2023-10-15 (CHB-3)

### Added

- Sending a command-like message gets its own workflow instead of storing it by [Arturo Vial](https://github.com/ratolibre1)

### Changed

- Renamed the module to "chat-hex" by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.3] - 2023-10-15 (CHB-2)

### Added

- Added new endpoint to list messages from a specific chatroom by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.2] - 2023-10-14 (CHB-1)

### Added

- Added Messages folders and endpoint to send a message by [Arturo Vial](https://github.com/ratolibre1)

## [0.0.1] - 2023-10-14

### Added

- Hexagonal Architecture scaffolding forked from [Github](https://github.com/favians/go-hexagonal) by [Arturo Vial](https://github.com/ratolibre1)

### Changed

- Cleaned some stuff, ensured connection with DB by [Arturo Vial](https://github.com/ratolibre1)


modules/preload/mongo_repo.go
package mongodb

import (
	"context"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type MongoDBRepository struct {
	usersCollection *mongo.Collection
	chatroomsCollection *mongo.Collection
}

func NewMongoDBRepository(db *mongo.Database) *MongoDBRepository {
	return &MongoDBRepository{
		db.Collection("users"),
		db.Collection("chatrooms"),
	}
}

func (repo *MongoDBRepository) PopulateMongoDB() error {
	clientOptions := options.Client().ApplyURI("mongodb://localhost:27017")
	client, err := mongo.Connect(context.TODO(), clientOptions)
	if err != nil {
		return err
	}
	defer client.Disconnect(context.TODO())

	usersData := []bson.M{
		{
			"_id":         "652a9d83f896448755f4440f",
			"email":       "artichoke@jobsity.com",
			"password":    "$2a$14$ukT1jGx.NRm/giDivYzlN.N7jQP/U8ErNwxICu.FiqTXky5wm5GUS",
			"name":        "Artichoke",
			"currentRoom": "room2",
		},
		{
			"_id":         "652a9dd9f896448755f44410",
			"email":       "lilbasil@jobsity.com",
			"password":    "$2a$14$ukT1jGx.NRm/giDivYzlN.N7jQP/U8ErNwxICu.FiqTXky5wm5GUS",
			"name":        "Lil' Basil",
			"currentRoom": "room1",
		},
		{
			"_id":         "653013ed4cca2b12068bf08f",
			"email":       "StockBot",
			"password":    "",
			"name":        "StockBot",
			"currentRoom": nil,
		},
	}

	chatroomsData := []bson.M{
		{
			"_id":   "652d548a4cca2b12068bf08d",
			"name":  "Offtopic",
			"desc":  "Come here and make some friends",
			"code":  "room1",
		},
		{
			"_id":   "652d557c4cca2b12068bf08e",
			"name":  "Serious Talk",
			"desc":  "Discuss technical stuff",
			"code":  "room2",
		},
	}

	usersCollection := repo.usersCollection
	var usersDataInterface []interface{}
	for _, user := range usersData {
		usersDataInterface = append(usersDataInterface, user)
	}
	_, err = usersCollection.InsertMany(context.TODO(), usersDataInterface)
	if err != nil {
		return err
	}

	chatroomsCollection := repo.chatroomsCollection
	var chatroomsDataInterface []interface{}
	for _, chatroom := range chatroomsData {
		chatroomsDataInterface = append(chatroomsDataInterface, chatroom)
	}
	_, err = chatroomsCollection.InsertMany(context.TODO(), chatroomsDataInterface)
	if err != nil {
		return err
	}
	
	return nil
}